### ISSUE TITLE
after alt-tab: endless phantom keypress eventtttttt

### ISSUE THREAD

**Issue by m1027** [labwc/labwc#1091]
Hi,

the issue I see is:

- First alt-tab to a console window (I am using alacritty in tmux)
- then the console receives keypress events (like mmmmmmm....) continously until I ctrl-c

More details:
 - The cursor glitches happen in a pure bash shall as well in a open vim file.
 - The phantom keyevent is not always the same. Maybe the last key I used in the terminal.
 - This happens apporx. once or twice a day.
 - This issue is present since approx. 3 months now.
 - My labwc version is mostly uptodate with git.

Do you have a hint what component could be responsible of this? Is it likely that labwc is concerned?

Thanks for awesome labwc.

**Comment by stefonarch** [comment]
It should happen with every application that accepts keys, what about other terminal emulators or text editors?

**Comment by heroin-moose** [comment]
It happens to me all the time :D

Electron apps, alacritty, firefox. You name it.

**Comment by Consolatis** [comment]
Hm.. This sounds like something that was fixed ages ago. Personally I did't experience any kind of wrong key-repeat since then.
Does it only happen in xwayland or wayland native applications or does it show up for both?

There is also the https://labwc.github.io/labwc-config.5.html#entry_keyboard_repeatdelay setting which defaults to 600 ms which might eventually be related.

**Comment by m1027** [comment]
@Consolatis: I am not using xwayland, just pure wayland. And I am using `<repeatDelay>300</repeatDelay>` and `<repeatRate>50</repeatRate>`.

**Comment by heroin-moose** [comment]
Native windows:
```
    <repeatRate>40</repeatRate>
    <repeatDelay>200</repeatDelay>
```

**Comment by kyak** [comment]
It happens to me as well. For some reason, the `l` key usually sticks.

This has something to do with swaylock, maybe. Because this happens sometimes after I unlock desktop, and I lock it by using `Super-l` combo (same `l` key that gets stuck).

Really hard to reproduce, it just happens.

**Comment by heroin-moose** [comment]
To me it mostly happens on layout switches.

**Comment by heroin-moose** [comment]
Yeah, I a way to reproduce!

1. Open a terminal
2. Press `s`
3. Without releasing `s` press right Alt to switch the layout
4. Release `s`

Works 10/10.

**Comment by johanmalm** [comment]
> To me it mostly happens on layout switches.

Layout? As in keyboard-layout.... or window-layout?

**Comment by heroin-moose** [comment]
Nah, its keybind or action, 100%.

**Comment by stefonarch** [comment]
> I found a way to reproduce!
> 
>     1. Open a terminal
> 
>     2. Press `s`
> 
>     3. Press any key binding
> 
>     4. Release `s`
> 
> 
> Works 10/10.
> 
> Some examples of "working" key bindings:
> 
> ```
>     <keybind key="ISO_Next_Group"><action name="Execute" command="true"/></keybind>
>     <keybind key="W-Space"><action name="ToggleAlwaysOnTop"/></keybind>
> ```

No way to reproduce it here.

**Comment by heroin-moose** [comment]
@stefonarch, do you have the latest master?

I can reproduce with this config:

```
<?xml version="1.0"?>
<labwc_config>
  <keyboard>
    <repeatRate>40</repeatRate>
    <repeatDelay>200</repeatDelay>
    <keybind key="W-Space"><action name="ToggleAlwaysOnTop"/></keybind>
  </keyboard>
</labwc_config>
```

and this command:

```
./src/labwc/build/labwc -s alacritty -c minimal.xml
```

**Comment by heroin-moose** [comment]
From a compositor perspective (I put `printf()` at `keyboard.c:438`):

```
00:00:01.334 [INFO] [../src/keyboard.c:438] start keybind repeat 31 1
00:00:02.449 [INFO] [../src/keyboard.c:438] start keybind repeat 125 1
00:00:02.557 [INFO] [../src/keyboard.c:438] start keybind repeat 57 1
00:00:02.639 [INFO] [../src/keyboard.c:438] start keybind repeat 57 0
00:00:02.707 [INFO] [../src/keyboard.c:438] start keybind repeat 125 0
00:00:02.722 [INFO] [../src/keyboard.c:438] start keybind repeat 31 0
```

I also can see that labwc does not cancel repeating:

```
void
keyboard_cancel_keybind_repeat(struct keyboard *keyboard)
{
	wlr_log(WLR_INFO, "in keyboard repeat");
	if (keyboard->keybind_repeat) {
		wlr_log(WLR_INFO, "cancelling keyboard repeat"); <--- not printed
		wl_event_source_remove(keyboard->keybind_repeat);
		keyboard->keybind_repeat = NULL;
	}
}
```

**Comment by heroin-moose** [comment]
Commenting this:

```
	/*
	 * If a press event was handled by a compositor binding, then do not
	 * forward the corresponding release event to clients
	 */
#if 0
	if (key_state_corresponding_press_event_was_bound(event->keycode)
			&& event->state == WL_KEYBOARD_KEY_STATE_RELEASED) {
		key_state_bound_key_remove(event->keycode);
		return true;
	}
#endif
```

Fixes things for me.

**Comment by heroin-moose** [comment]
So I guess the problem is that any key that was bound is considered to be a part of the last key binding handled by the compositor.

**Comment by heroin-moose** [comment]
Yep:

```
00:00:02.019 [INFO] [../src/keyboard.c:441] keybind key notify 31 1
00:00:03.452 [INFO] [../src/keyboard.c:441] start keybind repeat 125 1
00:00:03.589 [INFO] [../src/keyboard.c:441] start keybind repeat 57 1
00:00:03.695 [INFO] [../src/keyboard.c:441] start keybind repeat 57 0
00:00:03.695 [INFO] [../src/keyboard.c:242] removing bound key 57
00:00:03.699 [INFO] [../src/keyboard.c:441] start keybind repeat 125 0
00:00:03.699 [INFO] [../src/keyboard.c:242] removing bound key 125
00:00:03.718 [INFO] [../src/keyboard.c:441] start keybind repeat 31 0
00:00:03.718 [INFO] [../src/keyboard.c:242] removing bound key 31
```

**Comment by stefonarch** [comment]
> @stefonarch, do you have the latest master?
> 
> I can reproduce with this config:
> 
> ```
> <?xml version="1.0"?>
> <labwc_config>
>   <keyboard>
>     <keybind key="W-Space"><action name="ToggleAlwaysOnTop"/></keybind>
>   </keyboard>
> </labwc_config>
> ```
> 
> and this command:
> 
> ```
> $ ./build/labwc -s alacritty -C /var/empty -c minimal.xml
> ```

On master, yes. Using qterminal I can reproduce something similar, at every W-space it prints an s more and the action is not working. No idea where the s is coming from though

![screenshot_2023-09-23_12 35 50](https://github.com/labwc/labwc/assets/10681413/94a655ac-84bc-4ca7-82e0-492f47bafbb9)

**Comment by heroin-moose** [comment]
No, this is something else. Press s, wait for repeat to kick in, press W-space and only after that release s. You will see that autorepeat is not canceled.

**Comment by Consolatis** [comment]
These indeed seem like two separate issues that may or may not have the same root cause.

**Comment by m1027** [comment]
Glad to see that we can reproduce it always now.

This way I see even more strange behavior in a terminal (I have only one window open so that alt-tab actually does not switch the window):
- Pressing "s" (without releasing), then adding alt-tab, starts deleting (!) all "s". When releasing "s" and then alt-tab, the "s" phantom keyevent kicks in and produces s until ctrl-c is pressed.
- Trying the same with the "a" key: It behaves the same except for the "deleting" part. So: I get aaaaaaaaaaaa..., then with alt-tab no "a" event anymore (although "a" not released), then when releasing "a" and then alt-tab, I receive aaaaaaaaaaaaaa... without pressing it.
- Trying the same with the "l" key (small "L") behaves correctly! No phantom keyevents.

Huh!

**Comment by johanmalm** [comment]
I think it's important that surfaces do not receive release-events for keys where the corresponding press-event was absorbed by a keybind. I can't remember the detail, but this used to cause issues with dmenu_run (for example if bound to `W-d` you'd get dddddd... when dmenu_run opened).

The following commits refer
labwc/labwc@de99a8ba
labwc/labwc@7e57b7fc

And also issue #510

I think we should focus on making sure we use `key_state_corresponding_press_event_was_bound()` correctly and not include keys which are not part of the keybind.

**Comment by xxhh1112** [comment]
qqq

**Comment by m1027** [comment]
Thinking of the overall root cause:

Well, using two hands for typing, it may indeed happen to alt-tab (left hand) while still pressing another key (right hand). Thus, a keybind combination and the ordinary keystroke may indeed overlap... This seems to trigger the whole issue. At least for the issue I was addressing by this thread.

So, how to handle such case generally?

IMHO I wonder whether such overlapping double key combination should be considered valid at all. Why should alt-tab be triggered like an ordinary alt-tab although another key is pressed at the same time.

Without going into the code I am sorry that I cannot judge on how to solve this.

**Comment by johanmalm** [comment]
> IMHO I wonder whether such overlapping double key combination should be considered valid at all. Why should alt-tab be triggered like an ordinary alt-tab although another key is pressed at the same time.

I think you're right. We just need to figure out a way to implement it.

**Comment by heroin-moose** [comment]
> Why should alt-tab be triggered like an ordinary alt-tab although another key is pressed at the same time.

Because else keybindings like keyboard layout switching will not work during fast typing.

**Comment by Consolatis** [comment]
> > Why should alt-tab be triggered like an ordinary alt-tab although another key is pressed at the same time.
> 
> Because else keybindings like keyboard layout switching will not work during fast typing.

Changing the keyboard layout is not a keybind though. It is handled by libxkbcommon directly.

**Comment by heroin-moose** [comment]
The problem is that it still passed to the clients.

**Comment by heroin-moose** [comment]
Also is there a guarantee that release event will always be processes first?

**Comment by m1027** [comment]
> > Why should alt-tab be triggered like an ordinary alt-tab although another key is pressed at the same time.
> 
> Because else keybindings like keyboard layout switching will not work during fast typing.

Well... Fast typing is the only thing I had in mind, too, without mentioning. To not overcomplicate things. And because I cannot judge if this is a real problem or not.

Anyway, if we want to consider that: What's the big picture in this case then?

The main use case for a key handler is probably to just detect and handle keybindings properly. So, when I think of standard apps, a keybinding for ctrl-v (for pasting) is to be handled, and it is just not the same as ctrl-shift-v.

But of course we can add special cases to that when necessary. Here I am addressing  the "fast typing" case and alt-tab. If alt-tab + another key like "s" are overlapping key events, and we want to make them work like *separate* or sequential events then we need to handle that extra. Such special cases justify special handling in code. 

To implement this special case I imagine quite a lot of corner cases though:
- What if alt-tab-s is itself bound as another keybinding? Then it must not be separated into alt-tab plus something else.
- Order: If pressing "s" and then adding alt-tab at the same time, which causes the windows (thus the keystroke consumers) to switch, shouldn't the original window receive a phantom keyrelease event for "s" and the new window a phantom keypress event for "s"?
- Or the other way round: If pressing alt-tab first and, without releasing, adding "s" afterwards, there should not be such keyrelase event for the original window...

I hope that makes sence a bit and helps finding a good implementation.

I wonder how other WMs handle all this.

**Comment by johanmalm** [comment]
@m1027 Thanks for analysis. Did you see PR #1108? Testing welcome.

**Comment by m1027** [comment]
I tested PR #1108 against the above alt-tab + a test. Indeed that works now for me!

Thanks to everybody and the fast reaction on this!

I am closing this issue, feel free to reopen if there are regressions.

**Comment by Consolatis** [comment]
Leaving this open until the PR is merged.
